### PR TITLE
Convert Index and WriteBatch

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This converts the last bit of the client code to TS. There are still tests in JS.

I also enabled TSLint for the API test (test/typescript.ts), which uncovered a slew of problems.

### CODE REVIEW

File: dev/src/index.ts [hunk 1]
```diff
@@ -1121,77 +1174,69 @@ follow these steps, YOUR APP MAY BREAK.`);
...
- * @param {string} options.allowDeletes At what level field deletes are
- * supported (acceptable values are 'none', 'root' or 'all').
- * @param {boolean} options.allowServerTimestamps Whether server timestamps
- * are supported.
- * @param {number=} depth The current depth of the traversal.
- * @param {number=} inArray Whether we are inside an array.
- * @returns {boolean} 'true' when the object is valid.
- * @throws {Error} when the object is invalid.
+ * @param val JavaScript value to validate.
+ * @param  options Validation options
```
Comment (mikelehen) [inline dev/src/index.ts]: nit: Maybe grep the code for "@param  " because I know there's at least a couple of these.

File: dev/src/index.ts [hunk 2]
```diff
@@ -1290,15 +1330,7 @@ function validateDocumentData(obj, options) {
...
-/**
- * Sets the log function for all active Firestore instances.
- *
- * @method Firestore.setLogFunction
- * @param {Firestore~logFunction} logger - A log function that takes a single
- * string.
- */
-Firestore.setLogFunction = setLogFunction;
-
+// tslint:disable-next-line:no-default-export
```
Comment (mikelehen) [inline dev/src/index.ts]: It looks like you ended up adding this below...

File: dev/src/types.ts [hunk 3]
```diff
@@ -32,10 +33,57 @@ export type ApiMapValue = {
...
+   * `Timestamp` instead.
+   *
+   * NOTE: in the future `timestampsInSnapshots: true` will become the
+   * default and this option will be removed so you should change your code to
+   * use `Timestamp` now and opt-in to this new behavior as soon as you can.
+   */
+  timestampsInSnapshots?: boolean;
+
+  // tslint:disable-next-line:no-any
+  [key: string]: any;  // Accept other properties, such as GRPC settings.
```
Comment (mikelehen) [inline dev/src/types.ts]: By the way, what version of TypeScript are we using?  TypeScript 3.0 introduces the type-safe type `unknown`, which may be more suitable than `any` for some of these things...

File: dev/src/write-batch.ts [hunk 4]
```diff
@@ -76,21 +78,36 @@ export class WriteResult {
...
     return (
         this === other ||
         (is.instanceof(other, WriteResult) &&
          this._writeTime.isEqual(other._writeTime)));
   }
 }
 
+
+/** Helper type to manage the list of writes in a WriteBatch. */
+// TODO: Replace with api.IWrite
```
Comment (mikelehen) [inline dev/src/write-batch.ts]: Who should do this? When? Why is it not being done now?

File: dev/src/write-batch.ts [hunk 5]
```diff
@@ -565,11 +583,11 @@ export class WriteBatch {
  * Validates that the update data does not contain any ambiguous field
  * definitions (such as 'a.b' and 'a').
  *
- * @param {Map.<FieldPath, *>} data - An update map with field/value pairs.
- * @returns {boolean} 'true' if the input is a valid update map.
+ * @param data - An update map with field/value pairs.
```
Comment (mikelehen) [inline dev/src/write-batch.ts]: Is the `-` supposed to be there?

File: tsconfig.json [hunk 6]
```diff
@@ -7,17 +7,12 @@
     "noImplicitAny": false, // TODO: Remove when all source are migrated to TS
```
Comment (mikelehen) [inline tsconfig.json]: Are we here yet? Or no?

File: dev/src/document.ts [hunk 7]
```diff
@@ -163,7 +167,7 @@ export class DocumentSnapshot {
    * @param {Object} obj - The object to store in the DocumentSnapshot.
```
Comment (mikelehen) [inline dev/src/document.ts]: Are these type annotations still intentional?

Comment (schmidt-sebastian) [reply dev/src/document.ts]: I only want to keep them for external methods, as they are used in JSDoc. Removed from here.

FYI: document.ts still has room for a bunch of TS-related improvements.

File: dev/src/index.ts [hunk 8]
```diff
@@ -223,46 +281,10 @@ class Firestore {
...
-    this._clientPool = null;
-
-    /**
-     * Whether the initialization settings can still be changed by invoking
-     * `settings()`.
-     * @private
-     * @type {boolean}
-     */
-    this._settingsFrozen = false;
+    } as AnyDuringMigration);
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/src/index.ts]: Yes, that's the goal.

File: dev/src/index.ts [hunk 9]
```diff
@@ -291,7 +313,7 @@ class Firestore {
    *
    * @param {object} settings The settings to use for all Firestore operations.
    */
-  settings(settings) {
+  settings(settings: Settings) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/src/index.ts]: Yes to fixing this instance now, and turning this check on later.

File: dev/src/index.ts [hunk 10]
```diff
@@ -657,16 +693,18 @@ class Firestore {
    * @returns {Promise<Array.<DocumentSnapshot>>} A Promise that contains an array with
    * the resulting documents.
    */
-  getAll_(docRefs, requestTag, transactionId) {
+  getAll_(
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/src/index.ts]: Haha. This follows the following convention:

_ + name : Method is actually private
name + _: Method is "package-protected"

I am planning on eventually getting rid of this convention :)

File: dev/src/index.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply dev/src/index.ts]: Fixed 8 of them.

File: dev/src/index.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply dev/src/index.ts]: For some reason, the JSDoc comment itself triggers this warning, as well as the actual line that defines it.  I suspect it has to do with

```
@module {Firestore} @google-cloud/firestore
```

File: dev/src/types.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply dev/src/types.ts]: We are using TS 3 as of last week. I filed a bug for `unknown` (b/117464450).

File: dev/src/write-batch.ts (hunk 4 above)
Comment (schmidt-sebastian) [reply dev/src/write-batch.ts]: I'll do this shortly. Wanted to keep this PR nice and concise.

File: dev/src/write-batch.ts (hunk 5 above)
Comment (schmidt-sebastian) [reply dev/src/write-batch.ts]: Sometimes. Sometimes not :/ Removed here.

File: tsconfig.json (hunk 6 above)
Comment (schmidt-sebastian) [reply tsconfig.json]: Soon. I think we might have to migrate the tests.

**Pull request merged**